projects/optic/src/commands/capture/capture.ts

```diff
@@ -567,41 +543,78 @@ async function setup(filePath: string): Promise<string> {
   return await getCaptureStorage(resolvedPath);
 }
 
-async function startApp(
+type Bailout = { didBailout: boolean; promise: Promise<any> };
+function startApp(
   command: string,
-  dir: string,
-  readyEndpoint: string | undefined,
-  readyInterval: number,
-  readyTimeout: number,
-  targetUrl: string
-): Promise<ChildProcessWithoutNullStreams> {
+  dir: string
+): [ChildProcessWithoutNullStreams, Bailout] {
   const cmd = commandSplitter(command);
   const app = spawn(cmd.cmd, cmd.args, { detached: true, cwd: dir });
 
   app.stderr.on('data', (data) => {
     logger.error(data.toString());
   });
 
-  // TODO: this doesn't get caught by the calller but doesn't seem to leave the proxy hanging.
-  // lets find a better way.
-  app.on('exit', (code) => {
-    if (code! > 0) {
-      throw Error(`Unexpected exit from the server with exit code: ${code}`);
-    }
-  });
+  const bailout: Bailout = {
```

Comment:
We wrap the `app.on` with a promise and use that wherever we would usually wait for something (i.e. server ready OR requests in flight) 

This can then be `Promise.race` which will resolve or reject with the first resolve OR rejection